tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslSocketFactory.java

```diff
@@ -0,0 +1,108 @@
+import java.io.InputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import javax.net.ssl.HandshakeCompletedListener;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+
+final class InstrumentedSslSocketFactory extends SSLSocketFactory {
+
+    private final SSLSocketFactory delegate;
+    private final HandshakeCompletedListener listener;
+
+    InstrumentedSslSocketFactory(SSLSocketFactory delegate, TaggedMetricRegistry metrics, String name) {
+        this.delegate = delegate;
+        this.listener = newHandshakeListener(metrics, name, "socket");
+    }
+
+    @Override
+    public String[] getDefaultCipherSuites() {
+        return delegate.getDefaultCipherSuites();
+    }
+
+    @Override
+    public String[] getSupportedCipherSuites() {
+        return delegate.getSupportedCipherSuites();
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, String host, int port, boolean autoClose) throws IOException {
+        return wrap(delegate.createSocket(socket, host, port, autoClose));
+    }
+
+    @Override
+    public Socket createSocket(String host, int port) throws IOException {
+        return wrap(delegate.createSocket(host, port));
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress inetAddress, int clientPort)
+            throws IOException {
+        return wrap(delegate.createSocket(host, port, inetAddress, clientPort));
+    }
+
+    @Override
+    public Socket createSocket(InetAddress inetAddress, int port) throws IOException {
+        return wrap(delegate.createSocket(inetAddress, port));
+    }
+
+    @Override
+    public Socket createSocket(InetAddress inetAddress, int port, InetAddress clientAddress, int clientPort)
+            throws IOException {
+        return wrap(delegate.createSocket(inetAddress, port, clientAddress, clientPort));
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, InputStream consumed, boolean autoClose) throws IOException {
+        return wrap(delegate.createSocket(socket, consumed, autoClose));
+    }
+
+    @Override
+    public Socket createSocket() throws IOException {
+        return wrap(delegate.createSocket());
+    }
+
+    @Override
+    public String toString() {
+        return "InstrumentedSSLSocketFactory{delegate=" + delegate + '}';
+    }
+
+    private Socket wrap(Socket socket) {
+        if (socket instanceof SSLSocket) {
+            ((SSLSocket) socket).addHandshakeCompletedListener(listener);
+        }
+        return socket;
+    }
+
+    static HandshakeCompletedListener newHandshakeListener(
+            TaggedMetricRegistry metrics, String name, String type) {
+        return event -> metrics.meter(MetricName.builder()
+                .safeName("tls.handshake")
+                .putSafeTags("name", name)
+                .putSafeTags("cipher", event.getCipherSuite())
+                .putSafeTags("protocol", event.getSession().getProtocol())
```

Comment:
There was some discussion about tracking session resumption as well - is it possible to grab that info?

Reply:
Not easily, as far as I can tell. Since it's non-trivial I'd prefer to hold off for another PR to implement that feature.

Reply:
👍